### PR TITLE
HPCC-25744 Prometheus metrics sink

### PR DESCRIPTION
**Pull request by rpastrana**
Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@kenrowland take a look at a high level

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:224]: Appears you do the conversion of the type twice. What if the conversion routine just returned something like "custom" and you reported the value as such. That way you'd always report all values. Not sure makes sense.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:229]: Not sure if logging is appropriate or not. I feel that if you want metrics output in the log as well that you would load the log sink in addition to the prometheus sink.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:240]: Trivial, but toPrometheusMetric implies a single metric is handled when it's actually a vector.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:219]: You may be able to make this a const reference to avoid a copy.

**Review by kenrowland — COMMENTED** [review]
Please take a look at my comments and see if any warrant a change.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:229]: Agreed, this was mostly for my benefit during debugging. BTW is it possible to have multiple sinks concurrently?

**Comment by rpastrana** [comment]
@kenrowland made changes, please perform full review.

**Comment by rpastrana** [comment]
Hi @AttilaVamos I'm not sure what's going on w/ the checks on this PR, but I assume the failures are false negatives

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:220]: In all actuality, testing for this value being empty is not necessary, all metrics added to a reporter are required to have a unique name. The value will never be empty.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp]: The framework allows multiple sinks and treats them independently.

**Review by kenrowland — DISMISSED** [review]
Looks fine. I would make the one last change to remove the check for an empty metric name. I should have caught that last review. After that a merge is fine. I shall mark approved.

**Comment by AttilaVamos** [comment]
@rpastrana That error in teststdlibrary (tmod=DataPatterns) is a strange one and it caused by a "Disk Full" error. I saw there is a kind of test code as well. Does that using extensive disk space?
I rescheduled this PR to see whether that problem is persistent.

**Comment by rpastrana** [comment]
@AttilaVamos looks like the tests have passed now, thanks.

**Comment by rpastrana** [comment]
@kenrowland made the last change you recommended, but I also want to add the helm chart configuration schema for this sink in this PR

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (rpastrana) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:52]: @afishbeck just realized the createSecureSocketContextEx2 API operates on a "legacy" style config tree and wondered if we should be doing something special to tie into the containerized secrets management. 

sslConfig :
            cipherList:  !el "ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5"
            passphrase: !el someencodedpassphrase
            certificate: !el /home/hpcc/certificate/key.pem
            privatekey: !el /home/hpcc/certificate/public.key.pem
            verify:
            - enable: true
              address_match: true
              accept_selfsigned: true
              trusted_peers: !el a|b|c
              ca_certificates:
              - path: /home/hpcc/certificate

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:236]: This should be an internal error, caught earlier.  It shouldn't be possible to start the sink without a valid place to get metrics.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:253]: unused - should be caught by -Wall=error

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:271]: trivial  if (strieq()) is clearer to read.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:324]: I don't think you should be calling any function with esp in the name.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:283]: References to IEspContext seem wrong.  This code is not running within esp, and that looks like a significant chunk of functionality being pulled in.

File: system/metrics/sinks/prometheus/CMakeLists.txt
Comment (ghalliday) [inline system/metrics/sinks/prometheus/CMakeLists.txt:46]: I think esphttp is pulling in much too much code:

   ${XALAN_LIBRARIES}
    ${XERCES_LIBRARIES}
    jlib
    xmllib
    dalibase
    thorhelper      # this should not be dependent on the engine helper dll
    workunit        # neither should it be dependent on the wokunit dll
 
I think it will need a lighter weight library to process the requests.  @afishbeck any suggestions?

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:211]: This could potentially generate a lot of logging.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:221]: Should the #HELP only be output the first time?  Should that be an option?  Same for the # TYPE

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:243]: minor: Clearer without the DEFAULT_ prefix - unless it can be changed.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusReporterService.cpp:220]: Should be a const & otherwise it clones the string each time

**Review by ghalliday — CHANGES_REQUESTED** [review]
@rpastrana some initial comments.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (kenrowland) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:75]: @ghalliday One of the reasons I hold off on giving the reporter to the sink until collection starts is that sink init takes place before the component has fully added all metrics. It was a way to enforce preventing the sink from accessing the reporter until all init had completed. If we don't want to enforce that, we can make the change, however, sinks should know to wait until collection is started before calling the reporter.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:52]: Agreed

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:32]: Actually it is because it is called from another class that saves it.  I added a separate question about why there are 2 classes.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (ghalliday) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:75]: @kenrowland We support dynamic metrics - so the sinks already need to support metrics being added after startCollection is called.  I can't think of any reason why they would call the reporter before startCollection() is called, and that can be a requirement if necessary.

Comment (kenrowland) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:75]: @ghalliday I agree that sinks must support metrics being added dynamically. I was looking for a way to enforce no sink calling the reporter until it had finished with its initialization. If a requirement on the sink is sufficient, then let me know and I will open a Jira to address it. I suggest we wait until after this PR is merged so that all sinks and the framework are handled in the same PR.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:211]: I'm conflicted about this one...
@kenrowland is there a way for the sink to query the framework for metrics to be reported at init time?
I know metrics can be added dynamically, but this would allow the sink to report potential problems once at init time, with out needed to keep track of metrics, and performing look-ups at runtime.

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:221]: Agreed, it seems wasteful to report it every time, however their examples and literature seem to indicate the help is expected on every output (if available). The sink won't keep track of the unique requester(s) nor sessions, and therefore only reporting it once might not be particularly helpful. Perhaps expose a 'verbose' option to expose/suppress the help?

File: system/metrics/sinks/prometheus/prometheusReporterService.cpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:283]: Agreed, unfortunately the CHttpRequest/CHttpResponse utilize that object to keep some generic, and some ESP specific context info - perhaps a bit of an overkill. It feels like there should be base httpcontext, extended by the IEspContext, but as is the ESPContext is referenced in the http req/resp logic.
This particular service def doesn't need to create the "httpSecureContext" and will pass nullptr instead.

Comment (ghalliday) [reply system/metrics/sinks/prometheus/prometheusReporterService.cpp:283]: @afishbeck what is the best solution for this problem.  You added system/httplib for the Hashicorp vault secrets.  Should that be used?  Should we compile that into jlib somehow so that the code is shared?

File: system/metrics/sinks/prometheus/CMakeLists.txt
Comment (ghalliday) [reply system/metrics/sinks/prometheus/CMakeLists.txt:46]: Note, dafilesrv is one of the components that is likely to have metrics added.  It should be lightweight, and it would not pull in workunit, xmllib or xalan/xerces.

Comment (kenrowland) [reply system/metrics/sinks/prometheus/CMakeLists.txt:46]: Bringing in esphttp will include some of the base ESP metrics such as active request count. That would mean all components would get an ESP related active request count as part of their metrics as the metric is created statically.

File: system/metrics/sinks/prometheus/prometheusSink.cpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusSink.cpp:35]: It seemed like a reasonable logic abstraction layer... will any other component ever reference either class individually?? maybe not.

File: system/metrics/testing/metrictest.cpp
Comment (kenrowland) [reply system/metrics/testing/metrictest.cpp:76]: I will start looking at converting the test program to a set of unit tests for the reporter and for each individual sink. Eventually the test program should be removed.

**Comment by rpastrana** [comment]
@ghalliday please review httplib rendition

File: system/metrics/sinks/prometheus/prometheusSink.hpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusSink.hpp:41]: lines 41-46 look like they are duplicates of 34-39.

File: system/metrics/sinks/prometheus/prometheusSink.hpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusSink.hpp:84]: I don't think this is a legal c++ string.  I'm not sure why the compiler is accepting it.  You could either use string concatenation:

const char * s = "abc"
    "def\n"
    "...";

or use the raw string syntax if you want embedded newlines, and no need to quote ".

const char * s = R"!!(<html>
...
)!!";

File: system/metrics/sinks/prometheus/prometheusSink.hpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusSink.hpp:66]: trivial - not quite sure why this isn't with the other strings (or inlined since only used once)

**Review by ghalliday — CHANGES_REQUESTED** [review]
@rpastrana looks very close to ready.  A few minor comments.

**Comment by rpastrana** [comment]
@ghalliday please confirm lastest changes and I'll squash

File: system/metrics/sinks/prometheus/prometheusSink.hpp
Comment (ghalliday) [inline system/metrics/sinks/prometheus/prometheusSink.hpp:78]: Have you tested this?  I don't think you want the \ in front of the "s.  Also, will it be a problem having a newline before \< html >?

**Review by ghalliday — CHANGES_REQUESTED** [review]
@rpastrana

File: system/metrics/sinks/prometheus/prometheusSink.hpp
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusSink.hpp:78]: I passed my eye-ball test: 
image
but you're right, I'll update the string.

**Comment by rpastrana** [comment]
@ghalliday corrected the escape chars, and added proper html doctype declaration.

**Review by ghalliday — APPROVED** [review]
@rpastrana please squash and I will mrge

**Comment by rpastrana** [comment]
@ghalliday squashed.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**
